node/src/components/linear_chain_sync/operations.rs

```diff
@@ -473,17 +478,26 @@ async fn fetch_and_store_block_by_hash(
     }
 }
 
-/// Synchronize the trie store under a given state root hash.
+/// Synchronizes the trie store under a given state root hash.
 async fn sync_trie_store(
     effect_builder: EffectBuilder<JoinerEvent>,
     state_root_hash: Digest,
+    max_parallel_trie_fetches: usize,
 ) -> Result<(), LinearChainSyncError> {
     info!(?state_root_hash, "syncing trie store",);
+    // TODO: This implementation works like a stream with ordered buffering. Use an actual stream
+    //       with unordered buffering instead.
     let mut outstanding_trie_keys = vec![state_root_hash];
-    while let Some(trie_key) = outstanding_trie_keys.pop() {
-        let missing_descendant_trie_keys =
-            fetch_trie_and_insert_into_trie_store(effect_builder, trie_key).await?;
-        outstanding_trie_keys.extend(missing_descendant_trie_keys);
+    let mut fetches_in_progress = VecDeque::with_capacity(max_parallel_trie_fetches);
+    while !fetches_in_progress.is_empty() || !outstanding_trie_keys.is_empty() {
+        let fetches_to_start = max_parallel_trie_fetches.saturating_sub(fetches_in_progress.len());
+        let new_fetches = outstanding_trie_keys
+            .drain(outstanding_trie_keys.len().saturating_sub(fetches_to_start)..)
+            .map(|trie_key| tokio::spawn(fetch_and_store_trie(effect_builder, trie_key)));
+        fetches_in_progress.extend(new_fetches);
+        if let Some(join_handle) = fetches_in_progress.pop_front() {
+            outstanding_trie_keys.extend(join_handle.await??);
+        }
```

Comment:
Hmm, so if the fetches that are later in `fetches_in_progress` finish earlier, we won't reuse their slots until the first one finishes, right?
Is there maybe a way of awaiting until _one of_ the futures completes? So build something like a thread pool... I'd expect there should be one, but I don't know. `select!` looks close to it, but not quite (I think it cancels the futures other than the first one to complete?).

Reply:
That's exactly as far as I got, too: `select!` cancels the other futures, so we can't use it.

I know this approach is a bit crude, but at least it's not much code.